src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs

```diff
@@ -56,30 +61,60 @@ public CSharpType(Type type, bool isNullable, IReadOnlyList<CSharpType> argument
             IsValueType = type.IsValueType;
             IsEnum = type.IsEnum;
             IsPublic = type.IsPublic && arguments.All(t => t.IsPublic);
+
+            #region Assign the attributes of this type
+            IsIEnumerable = type == typeof(IEnumerable);
+            IsIEnumerableOfT = type == typeof(IEnumerable<>);
+            IsIAsyncEnumerableOfT = type == typeof(IAsyncEnumerable<>);
+            IsReadOnlyDictionary = type == typeof(IReadOnlyDictionary<,>);
+            IsReadWriteDictionary = type == typeof(IDictionary<,>);
+            IsDictionary = IsReadOnlyDictionary || IsReadWriteDictionary;
+            IsReadOnlyMemory = type == typeof(ReadOnlyMemory<>);
+            IsReadWriteList = type == typeof(IList<>) || type == typeof(ICollection<>) || type == typeof(List<>);
+            IsReadOnlyList = type == typeof(IEnumerable<>) || type == typeof(IReadOnlyList<>);
+            IsList = IsReadOnlyList || IsReadWriteList || IsReadOnlyMemory;
+            IsCollection = IsDictionary || IsList;
+            IsArray = type.IsArray;
+
+            IsTask = type == typeof(Task);
+            IsTaskOfT = type == typeof(Task<>);
+            // TODO -- these are azure-specific. We might need to generalize these using ApiTypes
+            IsResponse = type == typeof(Response);
+            IsResponseOfT = type == typeof(Response<>);
+            IsPageable = type == typeof(Pageable<>);
+            IsAsyncPageable = type == typeof(AsyncPageable<>);
+            IsOperation = type == typeof(Operation);
+            IsOperationOfT = type == typeof(Operation<>);
+            IsOperationOfAsyncPageable = IsOperationOfT && arguments.Count == 1 && arguments[0].IsAsyncPageable;
+            IsOperationOfPageable = IsOperationOfT && arguments.Count == 1 && arguments[0].IsPageable;
+            #endregion
```

Comment:
I do not have preference but Alexander usually calls out we should not abuse the lazy-loaded property, and assign everything to them when constructed.

This is usually valuable for immutable types such as this, I read something related on this topic as a comparison between C++ and rust.
It is the best practice to avoid funny issues that we always initialize everything for a type when constructing it especially when it is imuttable